server/chat-plugins/the-studio.ts

```diff
@@ -0,0 +1,613 @@
+	},
+	trackhelp: [
+		`/track [song name], [artist] - Displays the most relevant search result to the song name (and artist if specified) provided.`,
+	],
+
+	addrec: 'addrecommendation',
+	addrecommendation(target, room, user) {
+		room = this.requireRoom();
+		Recs.checkRoom(room);
+		this.checkCan('show', null, room);
+		const [artist, title, url, description, ...tags] = target.split('|').map(x => x.trim());
+		if (!(artist && title && url && description && tags?.length)) {
+			return this.parse(`/help addrecommendation`);
+		}
+
+		const cleansedTags = tags.map(x => x.trim());
+		Recs.add(artist, title, url, description, user.name, cleansedTags, String(user.avatar));
+
+		this.privateModAction(`${user.name} added a recommendation for '${title}' by ${artist}.`);
+		this.modlog(`RECOMMENDATION`, null, `add: '${toID(title)}' by ${toID(artist)}`);
+	},
+	addrecommendationhelp: [
+		`/addrecommendation artist | song title | url | description | tag1 | tag2 | ... - Adds a song recommendation. Requires: + % @ * # &`,
+	],
+
+	delrec: 'removerecommendation',
+	removerecommendation(target, room, user) {
+		const targetRoom = Rooms.search('thestudio') || room;
+		Recs.checkRoom(targetRoom);
+		this.room = targetRoom;
+		const [artist, title] = target.split(`|`).map(x => x.trim());
+		if (!(artist && title)) return this.parse(`/help removerecommendation`);
+		const rec = Recs.get(artist, title);
+		if (!rec) throw new Chat.ErrorMessage(`Recommendation not found.`);
+		if (toID(rec.userData.name) !== user.id) {
+			this.checkCan('mute', null, targetRoom!);
+		}
+
+		Recs.delete(artist, title);
+
+		this.privateModAction(`${user.name} removed a recommendation for '${title}' by ${artist}.`);
+		this.modlog(`RECOMMENDATION`, null, `remove: '${toID(title)}' by ${toID(artist)}`);
+	},
+	removerecommendationhelp: [
+		`/removerecommendation artist | song title - Removes a song recommendation. Requires: % @ * # &`,
+		`If you added a recommendation, you can remove it on your own without being one of the required ranks.`,
+	],
+
+	suggestrec: 'suggestrecommendation',
+	async suggestrecommendation(target, room, user) {
+		room = this.requireRoom();
+		Recs.checkRoom(room);
+		if (!target) {
+			return this.parse('/help suggestrecommendation');
+		}
+		this.checkChat(target);
+		const [artist, title, url, description, ...tags] = target.split('|').map(x => x.trim());
+		if (!(artist && title && url && description && tags?.length)) {
+			return this.parse(`/help suggestrecommendation`);
+		}
+
+		const cleansedTags = tags.map(x => x.trim());
+		Recs.suggest(artist, title, url, description, user.name, cleansedTags, String(user.avatar));
+		this.sendReply(`Your suggestion for '${title}' by ${artist} has been submitted.`);
+		room.sendRankedUsers(
+			`|html|${await Recs.render({artist, title, url, description, userData: {name: user.name, avatar: String(user.avatar)}, tags: cleansedTags, likes: 0}, true)}`,
+			'%'
+		);
```

Comment:
```suggestion
		const html = await Recs.render({artist, title, url, description, userData: {name: user.name, avatar: String(user.avatar)}, tags: cleansedTags, likes: 0}, true)
		room.sendRankedUsers(`|html|${html}`, '%');
```
Slightly cleaner.

Reply:
Your suggestion breaks line length rules.

```suggestion
		const html = await Recs.render({
			artist, title, url, description,
			userData: {name: user.name, avatar: String(user.avatar)},
			tags: cleansedTags, likes: 0,
		}, true)
		room.sendRankedUsers(`|html|${html}`, '%');
```